### PR TITLE
C#: Choose between .NET framework or core DLLs in standalone

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR 
- fixes the lookup path of the Mono lib folder,
- adds logic to choose between .NET framework/core DLLs ,
- adds integration tests for standalone extraction showing the extracted references.

The changes help with the extraction of projects that target the full framework. For example:

```diff
Results for rapPayne/WebGoat.Net
 Query CallTargets.ql:
-  Result count - both/traced/standalone: 894/459/36
-  Result count in the matching 87 files - both/traced/standalone: 894/459/36
-   Matching 64.36%
+  Result count - both/traced/standalone: 1353/0/0
+  Result count in the matching 87 files - both/traced/standalone: 1353/0/0
+   Matching 100.0%

Results for jerryhoff/WebGoat.NET
 Query CallTargets.ql:
-  Result count - both/traced/standalone: 989/737/1685
-  Result count in the matching 156 files - both/traced/standalone: 989/737/25
-   Matching 56.48%
+  Result count - both/traced/standalone: 1654/72/1887
+  Result count in the matching 156 files - both/traced/standalone: 1654/72/0
+   Matching 95.83%
```

### CODE REVIEW

**Comment by tamasvajk** [comment]
DCA results don't look great, I checked a couple of the projects, and compiler errors were reported because of conflicting framework DLLs. The conflict was between nuget restored reference assemblies and installed DLLs. So I'll handle this todo: https://github.com/github/codeql/pull/14368/files#diff-0179d8d74e2e78735e1d6cde08b1cff43d2a132f2c53a86c277cfdf3bbd95017R102-R104

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 1]
```diff
@@ -111,6 +96,25 @@ public DependencyManager(string srcDir, IDependencyOptions options, ILogger logg
...
+            var existsNetCoreRefNugetPackage = false;
+            var existsNetFrameworkRefNugetPackage = false;
+
+            // Find DLLs in the .Net / Asp.Net Framework
+            if (options.ScanNetFrameworkDlls)
+            {
+                existsNetCoreRefNugetPackage = IsNugetPackageAvailable("microsoft.netcore.app.ref");
+                existsNetFrameworkRefNugetPackage = IsNugetPackageAvailable("microsoft.netframework.referenceassemblies");
+
+                if (existsNetCoreRefNugetPackage || existsNetFrameworkRefNugetPackage)
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:108]: The DCA changes show that this logic is not working great. If dotnet core 6 is installed, and there are net6.0 and net5.0 targets, then the reference assemblies of net5.0 are downloaded and preferred over the installed dotnet 6 DLLs. And this causes an issue if there are other references which are compiled against dotnet 6.0.

We could probably force the download of the ref assemblies for all target platforms, and we should choose the latest one from those.

File: csharp/ql/integration-tests/all-platforms/standalone_dependencies/Assemblies.expected [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/System.Xml.XPath.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/System.Xml.XmlDocument.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/System.Xml.XmlSerializer.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/System.Xml.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/System.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/WindowsBase.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/mscorlib.dll |
+| /microsoft.netcore.app.ref/6.0.13/ref/net6.0/netstandard.dll |
+| /netstandard.library/2.0.3/build/netstandard2.0/ref/System.ComponentModel.Composition.dll |
+| /newtonsoft.json/12.0.1/lib/net45/Newtonsoft.Json.dll |
```
Comment (tamasvajk) [inline csharp/ql/integration-tests/all-platforms/standalone_dependencies/Assemblies.expected:170]: There are multiple `Newtonsoft.Json.dll` variants in this Nuget package. They have the same name and version. Our logic for choosing one of them is not stable, so it's not working for integration test purposes.

File: csharp/ql/integration-tests/all-platforms/standalone_dependencies/test.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import os
+from create_database_utils import *
+
+os.environ['PROCESSOR_ARCHITECTURE'] = 'x64'
```
Comment (tamasvajk) [inline csharp/ql/integration-tests/all-platforms/standalone_dependencies/test.py:4]: This should not be merged. I added a fix for this in the internal repo.

File: csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/DB-CHECK.expected [hunk 4]
```diff
@@ -0,0 +1,5 @@
+[VALUE_NOT_IN_TYPE] predicate attributes(@attribute id, int kind, @type_or_ref type_id, @attributable target): Value 25 of field target is not in type @attributable. The value is however in the following types: @void_type. Appears in tuple (949897,0,960,25)
```
Comment (tamasvajk) [inline csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/DB-CHECK.expected:1]: This is being fixed in https://github.com/github/codeql/pull/14436

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs [hunk 5]
```diff
@@ -164,6 +185,12 @@ private void DoInitialize()
                                 implicitUsingNamespaces.Add(ns);
                             }
                         }
+
+                        // Determine project structure:
+                        isLegacyProjectStructureUsed |= MicrosoftCSharpTargets().IsMatch(line);
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs:190]:
```suggestion
                        isLegacyProjectStructureUsed = isLegacyProjectStructureUsed || MicrosoftCSharpTargets().IsMatch(line);
```

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs [hunk 6]
```diff
@@ -164,6 +185,12 @@ private void DoInitialize()
...
+                        isNewProjectStructureUsed |= ProjectSdk().IsMatch(line);
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs:191]:
```suggestion
                        isNewProjectStructureUsed = isNewProjectStructureUsed || ProjectSdk().IsMatch(line);
```

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs [hunk 7]
```diff
@@ -164,6 +185,12 @@ private void DoInitialize()
...
+                        isNewProjectStructureUsed |= FrameworkReference().IsMatch(line);
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs:192]:
```suggestion
                        isNewProjectStructureUsed = isNewProjectStructureUsed || FrameworkReference().IsMatch(line);
```

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs (hunk 5 above)
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs:190]: `|=` uses logical or `|`, which does not have short circuit semantics (and unfortunately there doesn't exist `||=`).
You are also more than welcome to change the logic for `useAspNetCoreDlls` as well.

Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FileContent.cs:190]: I'm leaving the multi-line variants, such as `useAspNetCoreDlls`, as they are.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Runtime.cs [hunk 8]
```diff
@@ -107,33 +106,23 @@ private static IEnumerable<string> DesktopRuntimes
...
+        /// <summary>
+        /// Gets the .NET Framework location. Either the installation folder on Windows or Mono
+        /// </summary>
+        public string? DesktopRuntime => DesktopRuntimes?.FirstOrDefault();
 
-            // A bad choice if it's the self-contained runtime distributed in codeql dist.
-            return ExecutingRuntime;
-        }
+        /// <summary>
+        /// Gets the executiing runtime location, this is the self contained runtime shipped in the CodeQL CLI bundle.
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Runtime.cs:119]:
```suggestion
        /// Gets the executing runtime location, this is the self contained runtime shipped in the CodeQL CLI bundle.
```

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Semmle.Extraction.CSharp.DependencyFetching.csproj [hunk 9]
```diff
@@ -8,6 +8,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <RuntimeIdentifiers>win-x64;linux-x64;osx-x64</RuntimeIdentifiers>
     <Nullable>enable</Nullable>
+    <NoWarn>$(NoWarn);CA1822</NoWarn>
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Semmle.Extraction.CSharp.DependencyFetching.csproj:11]: Why not make it static instead of disabling the warning?

File: csharp/extractor/Semmle.Extraction.Tests/FileContent.cs [hunk 10]
```diff
@@ -1,14 +1,15 @@
 using Xunit;
 using System.Collections.Generic;
 using Semmle.Extraction.CSharp.DependencyFetching;
+using System;
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.Tests/FileContent.cs:4]: nit: format usings (there is a VS code plugin recommendation for C#).

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Semmle.Extraction.CSharp.DependencyFetching.csproj (hunk 9 above)
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Semmle.Extraction.CSharp.DependencyFetching.csproj:11]: This is just stylistic. This way `AspNetCoreRuntime`, `ExecutingRuntime`, `DesktopRuntime`, and `NetCoreRuntime` can be called the same way.

File: csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/Assemblies.ql [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import csharp
+
+private string getPath(Assembly a) {
+  not a.getCompilation().getOutputAssembly() = a and
+  exists(string s | s = a.getFile().getAbsolutePath() |
+    result =
+      s.substring(s.indexOf("GitHub/packages/") + "GitHub/packages/".length() + 16, s.length())
```
Comment (michaelnebel) [inline csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/Assemblies.ql:7]: Why is `16` also added explicitly (looks like this is also the length of "GitHub/packages/"?

Comment (tamasvajk) [reply csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/Assemblies.ql:7]: We're using a 16-character-long hash as a subfolder name. Let me put these in constants...

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DotNet.cs [hunk 12]
```diff
@@ -37,12 +40,29 @@ private void Info()
...
+            {
+                // Ugly hack: we set the TargetFrameworkRootPath and NetCoreTargetingPackRoot properties to an empty folder:
+                var path = ".empty";
+                if (tempWorkingDirectory != null)
+                {
+                    path = Path.Combine(tempWorkingDirectory.ToString(), "emptyFakeDotnetRoot");
+                    Directory.CreateDirectory(path);
+                }
+
+                args += $" /p:TargetFrameworkRootPath=\"{path}\" /p:NetCoreTargetingPackRoot=\"{path}\"";
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DotNet.cs:57]: Looks fine to me :-D

File: csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/Assemblies.ql (hunk 11 above)
Comment (tamasvajk) [reply csharp/ql/integration-tests/all-platforms/standalone_dependencies_net48/Assemblies.ql:7]: https://github.com/github/codeql/blob/main/csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs#L283-L284

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 13]
```diff
@@ -111,6 +97,25 @@ public DependencyManager(string srcDir, IDependencyOptions options, ILogger logg
                 DownloadMissingPackages(allNonBinaryFiles);
             }
 
+            var existsNetCoreRefNugetPackage = false;
+            var existsNetFrameworkRefNugetPackage = false;
+
+            // Find DLLs in the .Net / Asp.Net Framework
+            if (options.ScanNetFrameworkDlls)
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:104]: Maybe add a comment that this section needs to be after the `UseNuGet` section as we now attempt to use framework dll's that might be downloaded as a part of the solution/project restore.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs [hunk 14]
```diff
@@ -154,38 +159,158 @@ public DependencyManager(string srcDir, IDependencyOptions options, ILogger logg
...
+                        continue;
+                    }
+                    var secondDirectorySeparatorCharIndex = lowerFilename.IndexOf(Path.DirectorySeparatorChar, firstDirectorySeparatorCharIndex + 1);
+                    if (secondDirectorySeparatorCharIndex == -1)
+                    {
+                        continue;
+                    }
+                    var subFolderIndex = secondDirectorySeparatorCharIndex + 1;
+                    var isInAnalyzersFolder = lowerFilename.IndexOf("analyzers", subFolderIndex) == subFolderIndex;
+                    if (isInAnalyzersFolder)
```
Comment (michaelnebel) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:217]: Why is this needed?

**Review by michaelnebel — APPROVED** [review]
Excellent work 👍

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs (hunk 14 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs:217]: Nuget packages can ship analyzers, code fixes, and source generators in the [analyzers](https://learn.microsoft.com/en-us/nuget/guides/analyzers-conventions) folder. These DLLs need to be passed in the `analyzers` csc argument and not in `references`. The base classes used in these DLLs are not part of the dotnet runtime (but the dotnet SDK), so if these are passed as references, they will produce compile errors, so we remove them from the reference list.
